src/tools/database/elasticsearch.ts

```diff
@@ -0,0 +1,198 @@
+import { z } from "zod";
+import { Client, ClientOptions } from "@elastic/elasticsearch";
+import {
+  CatIndicesResponse,
+  IndicesGetMappingResponse,
+  SearchRequest,
+  SearchResponse,
+  SearchHit,
+} from "@elastic/elasticsearch/lib/api/types.js";
+import { ValidationError } from "ajv";
+
+type ToolRunOptions = BaseToolRunOptions;
+
+export interface ElasticSearchToolOptions extends BaseToolOptions {
+  connection: ClientOptions;
+}
+
+export class ElasticSearchTool extends Tool<
+  JSONToolOutput<any>,
+  ElasticSearchToolOptions,
+  ToolRunOptions
+> {
+  name = "ElasticSearchTool";
+
+  description = `Can query data from an ElasticSearch database. IMPORTANT: strictly follow this order of actions:
+   1. LIST_INDICES - retrieve a list of available indices
+   2. GET_INDEX_DETAILS - get details of index fields
+   3. SEARCH - perform search or aggregation query on a specific index or pass the original user query without modifications if it's a valid JSON ElasticSearch query`;
+
+  inputSchema() {
+    return z.object({
+      action: z
+        .enum(["LIST_INDICES", "GET_INDEX_DETAILS", "SEARCH"])
+        .describe(
+          "The action to perform. LIST_INDICES lists all indices, GET_INDEX_DETAILS fetches details for a specified index, and SEARCH executes a search or aggregation query",
+        ),
+      indexName: z
+        .string()
+        .optional()
+        .describe("The name of the index to query, required for GET_INDEX_DETAILS and SEARCH"),
+      query: z
+        .string()
+        .optional()
+        .describe("Valid ElasticSearch JSON search or aggregation query for SEARCH action"),
+      start: z.coerce
+        .number()
+        .int()
+        .min(0)
+        .default(0)
+        .optional()
+        .describe(
+          "The record index from which the query will start. Increase by the size of the query to get the next page of results",
+        ),
+      size: z.coerce
+        .number()
+        .int()
+        .min(0)
+        .max(10)
+        .default(10)
+        .optional()
+        .describe("How many records will be retrieved from the ElasticSearch query. Maximum is 10"),
+    });
+  }
+
+  static {
+    this.register();
+  }
+
+  public constructor(options: ElasticSearchToolOptions) {
+    super(options);
+    if (!options.connection.cloud && !options.connection.node && !options.connection.nodes) {
+      throw new ValidationError([
+        {
+          message: "At least one of the properties must be provided",
+          propertyName: "connection.cloud, connection.node, connection.nodes",
+        },
+      ]);
+    }
+  }
+
+  @Cache()
+  protected get client(): Client {
+    try {
+      return new Client(this.options.connection);
+    } catch (error) {
+      throw new ToolError(`Unable to connect to ElasticSearch: ${error}`, [], {
+        isRetryable: false,
+        isFatal: true,
+      });
+    }
+  }
+
+  protected async _run(
+    input: ToolInput<this>,
+    _options?: ToolRunOptions,
```

Comment:
There is a third parameter, `run,` which you should use to get the `signal` form.